src/rust/aggregator/src/data/user_history.rs

```diff
@@ -348,70 +357,94 @@ impl Data for UserHistory {
-                "#,
-                x.size,
-                x.maker_order_id,
-                x.market_id,
-                x.time,
-            )
-            .execute(&mut transaction as &mut PgConnection)
-            .await
-            .map_err(|e| DataAggregationError::ProcessingError(anyhow!(e)))?;
-        }
-        if !fill_events.is_empty() {
-            for x in fill_events {
-                sqlx::query!(
-                    r#"
-                        INSERT INTO aggregator.aggregated_events VALUES (
-                            $1, $2
-                        );
-                    "#,
-                    x.txn_version,
-                    x.event_idx,
-                )
-                .execute(&mut transaction as &mut PgConnection)
-                .await
-                .map_err(|e| DataAggregationError::ProcessingError(anyhow!(e)))?;
+                mark_as_aggregated(&mut transaction, &x.txn_version, &x.event_idx).await?;
             }
         }
-        transaction.commit()
+        transaction
+            .commit()
             .await
             .map_err(|e| DataAggregationError::ProcessingError(anyhow!(e)))?;
         Ok(())
     }
 }
+
+async fn aggregate_fill<'a>(
+    tx: &mut Transaction<'a, Postgres>,
+    size: &BigDecimal,
+    maker_order_id: &BigDecimal,
+    market_id: &BigDecimal,
+    time: &DateTime<Utc>,
+) -> DataAggregationResult {
+    sqlx::query!(
+        r#"
+            UPDATE aggregator.user_history
+            SET
+                remaining_size = remaining_size - $1,
+                total_filled = total_filled + $1,
+                order_status = CASE remaining_size
+                    WHEN 0 THEN CASE order_status
+                        WHEN 'cancelled' THEN order_status
+                        ELSE 'closed'
+                    END
+                    ELSE CASE order_type
+                        WHEN 'swap' THEN 'closed'
+                        ELSE order_status
+                    END
+                END,
+                last_updated_at = $4
+            WHERE order_id = $2 AND market_id = $3;
+        "#,
+        size,
+        maker_order_id,
+        market_id,
+        time,
+    )
+    .execute(tx as &mut PgConnection)
+    .await
+    .map_err(|e| DataAggregationError::ProcessingError(anyhow!(e)))?;
+    Ok(())
+}
+
+async fn aggregate_change<'a>(
+    tx: &mut Transaction<'a, Postgres>,
+    new_size: &BigDecimal,
+    order_id: &BigDecimal,
+    market_id: &BigDecimal,
+    time: &DateTime<Utc>,
+) -> DataAggregationResult {
+    sqlx::query!(
+        r#"
+            UPDATE aggregator.user_history
+            SET remaining_size = $1, last_updated_at = $4
+            WHERE order_id = $2 AND market_id = $3;
+        "#,
+        new_size,
+        order_id,
+        market_id,
+        time,
+    )
+    .execute(tx as &mut PgConnection)
+    .await
+    .map_err(|e| DataAggregationError::ProcessingError(anyhow!(e)))?;
+    Ok(())
+}
+
```

Comment:
suggest to add `aggregate_cancel` so that the status will be correct.

Reply:
```rust
for x in &cancel_events {
    sqlx::query!(
        r#"
            UPDATE aggregator.user_history
            SET order_status = 'cancelled', last_updated_at = $3
            WHERE order_id = $1 AND market_id = $2;
        "#,
        x.order_id,
        x.market_id,
        x.time,
    )
    .execute(&mut transaction as &mut PgConnection)
    .await
    .map_err(|e| DataAggregationError::ProcessingError(anyhow!(e)))?;
}
if !cancel_events.is_empty() {
    for x in cancel_events {
        mark_as_aggregated(&mut transaction, &x.txn_version, &x.event_idx).await?;
    }
}
```

It is done just above the `mark_as_aggregate` part. This is not a function as it is used only once (in contrast with fills and changes).